### PR TITLE
Idempotent egress IP setup + cloud egress IP implementation 

### PR DESCRIPTION
**Pull request by alexanderConstantinescu**
This PR has two main changes:

- Refactoring the egress IP code to have a more atomic and idempotent update mechanism
- Implementing the required functionality for interfacing with OpenShift's [cloud-network-config-controller](https://github.com/openshift/cloud-network-config-controller/) and supporting egress IP on public cloud platforms. More about this has been detailed in [the OpenShift enhancement proposal](https://github.com/openshift/enhancements/pull/571). Reviewers are encouraged to read that proposal as to understand this PR better. 

Both of these changes were made as to support egress IP on public cloud platforms and for interfacing properly with the cloud-network-config-controller. 

Note: the commit `Adding platform-type config parameter` will also be required for implementing platform aware functionality, like PR: https://github.com/openshift/ovn-kubernetes/pull/811 is trying to do. 

**- What this PR does and why is it needed**

**- Special notes for reviewers**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: go-controller/pkg/util/ovn.go [hunk 1]
```diff
@@ -67,3 +69,9 @@ func GetDatapathUUID(datapathName string) (string, error) {
 	}
 	return datapath, nil
 }
+
+func PlatformTypeIsEgressIPCloudProvider() bool {
```
Comment (trozet) [inline go-controller/pkg/util/ovn.go:73]: so it looks like you use this function to determine if the new controller should be enabled, but this is seems incorrect to me. The platform type has nothing to do with if you want to enable this controller. That should be a separate config flag.

File: go-controller/pkg/factory/factory.go [hunk 2]
```diff
@@ -18,6 +18,10 @@ import (
 	egressipinformerfactory "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1/apis/informers/externalversions"
 	egressiplister "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1/apis/listers/egressip/v1"
 
+	ocpcloudnetworkapi "github.com/openshift/api/cloudnetwork/v1"
```
[the remark this answers is not in the document]
Comment (trozet) [reply go-controller/pkg/factory/factory.go:21]: Also this needs a doc commit explaining what this controller is and how it functions with ovn-kube. It needs to address things like:
1. can this only be used with openshift integration?
2. if not, how do you use it with upstream?
3. any alternatives to use egress IP on clouds without this controller?
4. how to enable this feature?
5. how does it integrate with ovnkube?
6. how does it work?

**Review by trozet — CHANGES_REQUESTED** [review]
generally looks good. I don't know the egressIP code enough to review this line by line, but the gist of moving to the level driven approach seems right to me. Some comments inline.

File: go-controller/pkg/util/ovn.go (hunk 1 above)
Comment (alexanderConstantinescu) [reply go-controller/pkg/util/ovn.go:73]: > The platform type has nothing to do with if you want to enable this controller. 

It's completely tied to it. On public clouds egress IP will never work without that controller.

File: go-controller/pkg/factory/factory.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply go-controller/pkg/factory/factory.go:21]: OK, I will add a commit message explaining the component a bit more. I will also file PR later on for some documentation on using upstream ovn-kube with this controller.

File: go-controller/pkg/config/config.go [hunk 3]
```diff
@@ -858,6 +860,12 @@ var K8sFlags = []cli.Flag{
 		Destination: &cliConfig.Kubernetes.HostNetworkNamespace,
 		Value:       Kubernetes.HostNetworkNamespace,
 	},
+	&cli.StringFlag{
+		Name:        "platform-type",
+		Usage:       "The cloud provider platform type ovn-kubernetes is deployed on",
```
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply go-controller/pkg/config/config.go:865]: I will be more explicit and mention what the values for the platform-type should be.

File: go-controller/pkg/ovn/egressip.go [hunk 4]
```diff
@@ -1125,34 +1121,6 @@ func (e *egressIPController) deleteEgressReroutePolicy(filterOption, egressIPNam
 	return nil
 }
 
-func (e *egressIPController) deleteLegacyEgressReroutePolicies() error {
```
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply go-controller/pkg/ovn/egressip.go:1034]: No, this has nothing to do with local/shared gateway mode. A year ago the API for logical router policies changed an started supporting multiple `nexthops`, when migrating to this version of the API we needed to clean up all old logical router policies that defined a single `nexthop`, this function did that. But we've know been using this version of the API for many releases and this function is not syncing anything anymore.

File: go-controller/pkg/ovn/egressip.go [hunk 5]
```diff
@@ -433,10 +433,6 @@ func (oc *Controller) isEgressNodeReachable(egressNode *kapi.Node) bool {
 }
 
 func (oc *Controller) syncEgressIPs(eIPs []interface{}) {
-	if err := oc.eIPC.deleteLegacyEgressReroutePolicies(); err != nil {
```
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply go-controller/pkg/ovn/egressip.go:212]: Sure, I will file another PR with a dedicated README about egress IP for ovn-kubernetes.

File: go-controller/pkg/ovn/ovn.go [hunk 6]
```diff
@@ -901,51 +898,81 @@ func (oc *Controller) WatchEgressNodes() {
...
-			eIP := obj.(*egressipv1.EgressIP).DeepCopy()
-			oc.eIPC.assignmentRetryMutex.Lock()
-			defer oc.eIPC.assignmentRetryMutex.Unlock()
-			if err := oc.addEgressIP(eIP); err != nil {
-				klog.Error(err)
-			}
-			if err := oc.updateEgressIPWithRetry(eIP); err != nil {
-				klog.Error(err)
+			eIP := obj.(*egressipv1.EgressIP)
+			if err := oc.reconcileEgressIP(nil, eIP); err != nil {
```
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply go-controller/pkg/ovn/ovn.go:939]: Yes.

File: go-controller/pkg/ovn/egressip.go [hunk 7]
```diff
@@ -37,135 +39,222 @@ type egressIPDialer interface {
...
+		statusToKeep = append(statusToKeep, statusToAdd...)
+	}
+	if len(statusToAdd) > 0 || (len(statusToRemove) > 0 && new != nil) {
+		if err := oc.updateEgressIPStatus(newEIP.Name, statusToKeep); err != nil {
+			return err
+		}
+	}
+	metrics.RecordEgressIPCount(getEgressIPAllocationTotalCount(oc.eIPC.allocator))
+
+	if len(ipsToAdd) == 0 &&
```
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply go-controller/pkg/ovn/egressip.go:128]: I am updating the code with a bunch of doc blocs now.

File: go-controller/pkg/ovn/egressip.go [hunk 8]
```diff
@@ -37,135 +39,222 @@ type egressIPDialer interface {
...
 var dialer egressIPDialer = &egressIPDial{}
 
-func (oc *Controller) addEgressIP(eIP *egressipv1.EgressIP) error {
-	// If the status is set at this point, then we know it's valid from syncEgressIP and we have no assignment to do.
-	// Just initialize all watchers (which should not re-create any already existing items in the OVN DB)
-	if len(eIP.Status.Items) == 0 {
-		if err := oc.assignEgressIPs(eIP); err != nil {
-			return fmt.Errorf("unable to assign egress IP: %s, error: %v", eIP.Name, err)
-		}
+func (oc *Controller) reconcileEgressIP(old, new *egressipv1.EgressIP) error {
```
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply go-controller/pkg/ovn/egressip.go:45]: Same as below, updating the PR with a bunch of doc blocs

File: go-controller/pkg/ovn/egressip.go [hunk 9]
```diff
@@ -204,6 +204,11 @@ func (oc *Controller) syncEgressIPs(eIPs []interface{}) {
 				klog.Error(err)
 				continue
 			}
+		} else {
+			for _, eIPStatus := range eIP.Status.Items {
```
[the remark this answers is not in the document]
Comment (alexanderConstantinescu) [reply go-controller/pkg/ovn/egressip.go]: Will do

File: go-controller/pkg/util/ovn.go (hunk 1 above)
Comment (trozet) [reply go-controller/pkg/util/ovn.go:73]: sure i agree its tied to it, but the platform you are using shouldn't decide if egress ip with this controller is used. However, I just realized we have config.OVNKubernetesFeature.EnableEgressIP to enable egressIP. So I guess user configuring that and the platform is enough to say ok the intention is to enable this behavior.

Would it be possible for someone to use egress IP without the controller? For example, what if someone reserved an IP on a cloud and assigned it to a node manually, and then labeled just that node as egress IP with that IP. That should work right? Obviously, no HA there.

Comment (alexanderConstantinescu) [reply go-controller/pkg/util/ovn.go:73]: > Would it be possible for someone to use egress IP without the controller? 

Yes, someone could do what you say, but that would involve _a lot_ of work. Given the normal operations of egress IP (specifically that they can move whenever the egress IP controller detects a node as "unavailable"), doing that manually without this controller would make that persons life very very miserable. 

> For example, what if someone reserved an IP on a cloud and assigned it to a node manually, and then labeled just that node as egress IP with that IP

Yes, theoretically that would work.

**Review by trozet — APPROVED** [review]

**Pull request merged**